diatheke/streams.py

```diff
@@ -0,0 +1,282 @@
+    if isinstance(reader, io.TextIOBase):
+        is_text = True
+    else:
+        # Otherwise we assume we have a byte reader
+        is_text = False
+
+    while True:
+        # Read the next chunk of data
+        data = reader.read(buffSize)
+        if (is_text and data == '') or (not is_text and data == b''):
+            # Reached the EOF
+            break
+
+        # Send the audio
+        if not stream.send_audio(data):
+            break
+
+    return stream.result()
+
+
+def write_TTS_audio(stream, writer):
+    """Deprecated. Use client.write_tts_audio() instead."""
+    # Check if we have a text writer
+    if isinstance(writer, io.TextIOBase):
+        is_text = True
+    else:
+        # Otherwise we assume we have a byte writer
+        is_text = False
+
+    for data in stream:
+        if is_text:
+            # Convert the text to a string before writing.
+            writer.write(str(data.audio))
+        else:
+            writer.write(data.audio)
+
+
+class TranscribeStream(object):
+    """TranscribeStream represents a bidirectional stream where audio is
+    sent to Diatheke and transcriptions are returned to the client. This
+    differs from an ASRStream in purpose - where ASRStream accepts audio
+    to continue a conversation with the system, TranscribeStream accepts
+    audio solely for the purpose of getting a transcription that the client
+    can use for any purpose (e.g., note taking, send a message, etc.)."""
+
+    def __init__(self, client_stub):
+        # The Python implementation of gRPC is a little different from
+        # other languages. Rather than having send and receive methods
+        # for the stream, the bidirectional stream takes an iterator or
+        # generator function, and returns an iterable. This is a little
+        # inconvenient for how we want our SDK to be structured, so we
+        # use our own stream class with some threading to make this work.
+
+        # Set up queues to transfer data
+        input = Queue(maxsize=1)
+        output = Queue(maxsize=1)
+        self._input_queue = input
+        self._output_queue = output
+
+        # Use a flag (and lock) to indicate when the stream is closed.
+        lock = threading.Lock()
+        doneFlag = False
+
+        def is_done():
+            nonlocal doneFlag
+            with lock:
+                if doneFlag:
+                    return True
+                return False
+        self._is_done = is_done
+
+        # Generator function to pass data from the input queue to the
+        # streaming method.
+        def wait_for_audio():
+            while not is_done():
+                # Wait for data from the queue, which will block until data
+                # is available.
+                request = input.get()
+                if request is None:
+                    # This is our sentinel to indicate that the generator
+                    # should finish, which will end the stream.
+                    return
```

Comment:
```suggestion
                    return None
```
[`return` vs `return None` should be the same but just for sake of readability, i prefer `return None`](https://stackoverflow.com/a/15300733)

Reply:
Good point. It aligns with [PEP8](https://peps.python.org/pep-0008/#programming-recommendations),

> Be consistent in return statements. Either all return statements in a function should return an expression, or none of them should. If any return statement returns an expression, any return statements where no value is returned should explicitly state this as return None, and an explicit return statement should be present at the end of the function (if reachable):

and we also do it in other typed programming languages, so I guess this make sense.